pkg/batchscheduler/volcano/volcano_scheduler.go

```diff
@@ -0,0 +1,298 @@
+*/
+
+package volcano
+
+import (
+	"fmt"
+	"github.com/golang/glog"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+
+	apiextensionsclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+
+	"volcano.sh/volcano/pkg/apis/scheduling/v1alpha2"
+	volcanoclient "volcano.sh/volcano/pkg/client/clientset/versioned"
+
+	"github.com/GoogleCloudPlatform/spark-on-k8s-operator/pkg/apis/sparkoperator.k8s.io/v1beta1"
+	"github.com/GoogleCloudPlatform/spark-on-k8s-operator/pkg/batchscheduler/interface"
+)
+
+const (
+	PodGroupName = "podgroups.scheduling.sigs.dev"
+)
+
+type VolcanoBatchScheduler struct {
+	extensionClient apiextensionsclient.Interface
+	volcanoClient   volcanoclient.Interface
+}
+
+func GetPluginName() string {
+	return "volcano"
+}
+
+func (v *VolcanoBatchScheduler) Name() string {
+	return GetPluginName()
+}
+
+func (v *VolcanoBatchScheduler) ShouldSchedule(app *v1beta1.SparkApplication) bool {
+
+	checkScheduler := func(scheduler *string, annotation map[string]string) bool {
+		if scheduler != nil && *scheduler == v.Name() {
+			//return false if it's been scheduled already.
+			if existed, ok := annotation[v1alpha2.GroupNameAnnotationKey]; ok {
+				if existed == v.getAppPodGroupName(app) {
+					return false
+				}
+			}
+			return true
+		}
+		return false
+	}
+
+	if app.Spec.Mode == v1beta1.ClientMode {
+		return checkScheduler(app.Spec.Executor.SchedulerName, app.Spec.Executor.Annotations)
+	}
+	if app.Spec.Mode == v1beta1.ClusterMode {
+		return checkScheduler(app.Spec.Executor.SchedulerName, app.Spec.Executor.Annotations) && checkScheduler(app.Spec.Driver.SchedulerName, app.Spec.Driver.Annotations)
+	}
+
+	glog.Warningf("Unsupported Spark application mode %s, abandon schedule via volcano.", app.Spec.Mode)
+	return false
+}
+
+func (v *VolcanoBatchScheduler) BeforeSubmitSparkApplication(app *v1beta1.SparkApplication) (*v1beta1.SparkApplication, error) {
+	newApp := app.DeepCopy()
+	if newApp.Spec.Executor.Annotations == nil {
+		newApp.Spec.Executor.Annotations = make(map[string]string)
+	}
+
+	if newApp.Spec.Driver.Annotations == nil {
+		newApp.Spec.Driver.Annotations = make(map[string]string)
+	}
+
+	if newApp.Spec.Mode == v1beta1.ClientMode {
+		return v.syncPodGroupInClientMode(newApp)
+	} else if newApp.Spec.Mode == v1beta1.ClusterMode {
+		return v.syncPodGroupInClusterMode(newApp)
+	}
+	return newApp, nil
+}
+
+func (v *VolcanoBatchScheduler) syncPodGroupInClientMode(app *v1beta1.SparkApplication) (*v1beta1.SparkApplication, error) {
+	//We only care about the executor pods in client mode
+	newApp := app.DeepCopy()
+	if _, ok := newApp.Spec.Executor.Annotations[v1alpha2.GroupNameAnnotationKey]; !ok {
+		//Only executor resource will be considered.
+		if err := v.syncPodGroup(newApp, *newApp.Spec.Executor.Instances, getExecutorRequestResource(app)); err == nil {
```

Comment:
Shouldn't `MinMember` be 1, not the total number of executors? If you require they all be scheduled, you can't start the application early, increasing overall execution time. This also doesn't allow fair scheduling between applications.

Reply:
+1, 1 is better for fair-share :)

Reply:
Sure.